File: services/crm-connector/adapters.go
// Адаптеры CRM (решение D1): единый интерфейс для amoCRM и Bitrix24.
//
// В mock/dev-режиме используется InMemoryAdapter. Реальные адаптеры (amoCRM, Bitrix24)
// реализуются в эпике E5 поверх REST API соответствующих CRM (TODO).
package main

import (
	"os"
	"sync"
)

type CRMAdapter interface {
	Name() string
	UpsertContact(c ContactIn) string
	UpsertLead(l LeadIn) string
	UpdateDealStage(dealID, stage string)
	CreateTask(t TaskIn) string
	AddNote(n NoteIn) string
	GetContactByPhone(phone string) (ContactOut, bool)
}

// InMemoryAdapter — заглушка для разработки/тестов без реального доступа к CRM.
type InMemoryAdapter struct {
	name     string
	mu       sync.Mutex
	contacts map[string]ContactOut
	byPhone  map[string]string
}

func NewInMemoryAdapter(name string) *InMemoryAdapter {
	return &InMemoryAdapter{name: name, contacts: map[string]ContactOut{}, byPhone: map[string]string{}}
}

func (a *InMemoryAdapter) Name() string { return a.name }

func (a *InMemoryAdapter) UpsertContact(c ContactIn) string {
	a.mu.Lock()
	defer a.mu.Unlock()
	if c.Phone != "" {
		if id, ok := a.byPhone[c.Phone]; ok {
			return id
		}
	}
	id := newID()
	a.contacts[id] = ContactOut{CRMContactID: id, Name: c.Name, Phone: c.Phone}
	if c.Phone != "" {
		a.byPhone[c.Phone] = id
	}
	return id
}

func (a *InMemoryAdapter) UpsertLead(l LeadIn) string {
	a.UpsertContact(l.Contact)
	return newID()
}

func (a *InMemoryAdapter) UpdateDealStage(string, string) {}

func (a *InMemoryAdapter) CreateTask(TaskIn) string { return newID() }

func (a *InMemoryAdapter) AddNote(NoteIn) string { return newID() }

func (a *InMemoryAdapter) GetContactByPhone(phone string) (ContactOut, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	id, ok := a.byPhone[phone]
	if !ok {
		return ContactOut{}, false
	}
	c, ok := a.contacts[id]
	return c, ok
}

// TODO(E5): AmoCRMAdapter поверх amoCRM REST API (OAuth 2.0).
// TODO(E5): Bitrix24Adapter поверх Bitrix24 REST (входящий вебхук), учесть коробку (on-prem).

func getAdapter() CRMAdapter {
	mode := env("APP_MODE", "mock")
	provider := env("CRM_PROVIDER", "amocrm")
	if mode == "mock" {
		return NewInMemoryAdapter("memory")
	}
	// До реализации реальных адаптеров используем in-memory с именем провайдера.
	return NewInMemoryAdapter(provider)
}

func env(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

File: services/crm-connector/app/adapters.py
"""Адаптеры CRM (решение D1): единый интерфейс для amoCRM и Bitrix24.

В mock/dev-режиме используется InMemoryAdapter. Реальные адаптеры (AmoCRMAdapter,
Bitrix24Adapter) реализуются в эпике E5 поверх REST API соответствующих CRM.
"""
from __future__ import annotations

import os
import uuid
from abc import ABC, abstractmethod

from .models import ContactIn, ContactOut, LeadIn, NoteIn, TaskIn


class BaseCRMAdapter(ABC):
    name: str = "base"

    @abstractmethod
    def upsert_contact(self, contact: ContactIn) -> str: ...

    @abstractmethod
    def upsert_lead(self, lead: LeadIn) -> str: ...

    @abstractmethod
    def update_deal_stage(self, deal_id: str, stage: str) -> None: ...

    @abstractmethod
    def create_task(self, task: TaskIn) -> str: ...

    @abstractmethod
    def add_note(self, note: NoteIn) -> str: ...

    @abstractmethod
    def get_contact_by_phone(self, phone: str) -> ContactOut | None: ...


class InMemoryAdapter(BaseCRMAdapter):
    """Заглушка для разработки/тестов без реального доступа к CRM."""

    name = "memory"

    def __init__(self) -> None:
        self._contacts: dict[str, ContactOut] = {}
        self._by_phone: dict[str, str] = {}

    def upsert_contact(self, contact: ContactIn) -> str:
        if contact.phone and contact.phone in self._by_phone:
            return self._by_phone[contact.phone]
        cid = str(uuid.uuid4())
        self._contacts[cid] = ContactOut(
            crm_contact_id=cid, name=contact.name, phone=contact.phone
        )
        if contact.phone:
            self._by_phone[contact.phone] = cid
        return cid

    def upsert_lead(self, lead: LeadIn) -> str:
        self.upsert_contact(lead.contact)
        return str(uuid.uuid4())

    def update_deal_stage(self, deal_id: str, stage: str) -> None:  # noqa: D401
        return None

    def create_task(self, task: TaskIn) -> str:
        return str(uuid.uuid4())

    def add_note(self, note: NoteIn) -> str:
        return str(uuid.uuid4())

    def get_contact_by_phone(self, phone: str) -> ContactOut | None:
        cid = self._by_phone.get(phone)
        return self._contacts.get(cid) if cid else None


class AmoCRMAdapter(InMemoryAdapter):
    """TODO(E5): реальная реализация поверх amoCRM REST API (OAuth 2.0)."""

    name = "amocrm"


class Bitrix24Adapter(InMemoryAdapter):
    """TODO(E5): реальная реализация поверх Bitrix24 REST (входящий вебхук)."""

    name = "bitrix24"


def get_adapter() -> BaseCRMAdapter:
    mode = os.getenv("APP_MODE", "mock")
    provider = os.getenv("CRM_PROVIDER", "amocrm")
    if mode == "mock":
        return InMemoryAdapter()
    if provider == "bitrix24":
        return Bitrix24Adapter()
    return AmoCRMAdapter()
